File: lintcode/validate-binary-search-tree.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/validate-binary-search-tree/

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

import sys
class Solution:
    def helper(self, root, min_val, max_val):
        if not root:
            return True

        if root.val > min_val and root.val < max_val:
            return self.helper(root.left, min_val, root.val) and self.helper(root.right, root.val, max_val)

        return False

    """
    @param root: The root of binary tree.
    @return: True if the binary tree is BST, or false
    """
    def isValidBST(self, root):
        # write your code here
        return self.helper(root, -sys.maxint, sys.maxint)

File: leetcode/validate-binary-search-tree.cc
// Question: https://leetcode.com/problems/validate-binary-search-tree/

#include "headers.h"

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return helper(root, LONG_MAX, LONG_MIN);
    }

private:
    bool helper(TreeNode* root, long max_val, long min_val) {
        if (!root) {
            return true;
        }

        if ((long)root->val >= max_val || (long)root->val <= min_val) {
            return false;
        }

        return helper(root->left, root->val, min_val) && helper(root->right, max_val, root->val);
    }
};


int main(int argc, char** argv) {
    TreeNode* root = new TreeNode(10);
    root->left = new TreeNode(5);
    root->right = new TreeNode(15);
    root->left->left = new TreeNode(1);
    root->right->left = new TreeNode(16);
    Solution s;
    cout << s.isValidBST(root);
}
